### PR TITLE
44 extend substrate class

### PR DESCRIPTION
**Pull request by aalbino2**
@hampusnasstrom take a look to it. Whatever you think should be in the Lab plugin, I can easily remove from this Community plugin

### CODE REVIEW

File: src/nomad_material_processing/__init__.py [hunk 1]
```diff
@@ -94,7 +96,7 @@ class Parallelepiped(Geometry):
         a_eln=ELNAnnotation(
             component=ELNComponentEnum.NumberEditQuantity,
             defaultDisplayUnit='millimeter',
-            label='Width (x)',
+            label='Width (x or a)',
```
Comment (hampusnasstrom) [inline src/nomad_material_processing/__init__.py:99]: What is the idea of adding a, b, and c? I would stick with one of the notations, this becomes quite messy.

File: src/nomad_material_processing/__init__.py [hunk 2]
```diff
@@ -104,7 +106,126 @@ class Parallelepiped(Geometry):
...
+        ),
+        unit='meter',
+    )
+    side = Quantity(
+        type=float,
+        description='The x and y dimensions of the parallelepiped.',
+        a_eln=ELNAnnotation(
+            component=ELNComponentEnum.NumberEditQuantity,
+            defaultDisplayUnit='millimeter',
+            label='Side (x or a = y or b)',
```
Comment (hampusnasstrom) [inline src/nomad_material_processing/__init__.py:176]: This for example becomes very unclear.

File: src/nomad_material_processing/__init__.py [hunk 3]
```diff
@@ -80,11 +82,11 @@ class Parallelepiped(Geometry):
     m_def = Section()
     height = Quantity(
         type=float,
-        description='The z dimension of the parallelepiped.',
+        description='The z dimension (or thickness) of the parallelepiped.',
```
Comment (hampusnasstrom) [inline src/nomad_material_processing/__init__.py:85]: Does this mean that this is always the smallest dimension? If we have a thin cubic rod for example, the height could be the largest and if you asked people about the thickness of the rod they would probably give you the x or the y. I guess the same can be said for the length though...

File: src/nomad_material_processing/__init__.py [hunk 4]
```diff
@@ -123,6 +244,154 @@ class Parallelepiped(Geometry):
...
+    )
+
+
+class Irregular(Geometry):
+    """
+    A shape that does not fit into any of the other geometry classes.
+    """
+
+    m_def = Section()
+    height = Quantity(
```
Comment (hampusnasstrom) [inline src/nomad_material_processing/__init__.py:383]: A height would imply that you have to parallel surfaces though. Maybe we name it something like `ParallelSurfaces`?

File: src/nomad_material_processing/__init__.py [hunk 5]
```diff
@@ -177,6 +446,15 @@ class Dopant(ElementalComposition):
         ),
         unit='1 / m ** 3',
     )
+    doping_deviation = Quantity(
+        type=float,
+        description='The ± uncertainty in the doping level.',
```
Comment (hampusnasstrom) [inline src/nomad_material_processing/__init__.py:451]: If we call the name `_deviation` I would use "deviation" in the description as well.
```suggestion
        description='The ± deviation in the doping level.',
```
Do we need to specify the deviation from what?

File: src/nomad_material_processing/__init__.py [hunk 6]
```diff
@@ -209,6 +505,32 @@ class SubstrateCrystalProperties(CrystalProperties):
         Section describing any miscut of the substrate with respect to the substrate
         orientation.
         """,
+        repeats=True,
+    )
+
+
+class ElectricProperties(ArchiveSection):
```
Comment (hampusnasstrom) [inline src/nomad_material_processing/__init__.py:512]: I would call it `ElectricalProperties` or `ElectronicProperties`, the later being what we use in the NOMAD properties taxonomy.

File: src/nomad_material_processing/__init__.py [hunk 7]
```diff
@@ -209,6 +505,32 @@ class SubstrateCrystalProperties(CrystalProperties):
...
+    """
+    The electrical properties of a material.
+    """
+
+    m_def = Section()
+
+    conductivity_type = Quantity(
```
Comment (hampusnasstrom) [inline src/nomad_material_processing/__init__.py:519]: Not sure about this name. This property is actually missing from the NOMAD properties taxonomy. Conductivity type sounds more like whether it is thermal, electric, or ionic. I know we're in the ElectronicProperties section, but still. Was this what Oliver called it? I would talk about the majority carrier type but that would map to holes or electrons... Maybe semiconductor_type? I think we would need to discuss this with a PI.

File: src/nomad_material_processing/__init__.py [hunk 8]
```diff
@@ -209,6 +505,35 @@ class SubstrateCrystalProperties(CrystalProperties):
...
+        ),
+    )
+    carrier_density = Quantity(
+        type=np.dtype(float),
+        unit='1 / cm**3',
+        shape=['*'],
+        a_eln=ELNAnnotation(
+            component=ELNComponentEnum.NumberEditQuantity,
+        ),
+        description='Concentration of free charge carriers, electrons in the conduction band and holes in the valence band.',
```
Comment (hampusnasstrom) [inline src/nomad_material_processing/__init__.py:536]: How do you store this as an array? Is it a tuple of length 2 where the first element is electron concentration and the second is hole? Or is it a function of energy? or k?

**Review by hampusnasstrom — COMMENTED** [review]
Nice work with all the shapes! Generally it looks good but it did raise some questions.

File: src/nomad_material_processing/__init__.py [hunk 9]
```diff
@@ -219,15 +544,21 @@ class Substrate(CompositeSystem):
     """
 
     m_def = Section()
-    supplier = Quantity(
```
[the remark this answers is not in the document]
Comment (aalbino2) [reply src/nomad_material_processing/__init__.py:222]: changed back

File: src/nomad_material_processing/__init__.py (hunk 1 above)
Comment (aalbino2) [reply src/nomad_material_processing/__init__.py:99]: switched back to x-y-z

File: src/nomad_material_processing/__init__.py (hunk 8 above)
Comment (aalbino2) [reply src/nomad_material_processing/__init__.py:536]: solved

File: src/nomad_material_processing/__init__.py [hunk 10]
```diff
@@ -123,6 +244,154 @@ class Parallelepiped(Geometry):
     )
 
 
+class TruncatedCone(Geometry):
+    """
+    Class autogenerated from yaml schema.
```
Comment (hampusnasstrom) [inline src/nomad_material_processing/__init__.py:249]: Just saw this... I'll just make a suggestion here that you can accept if you want.
```suggestion
    A cone with the top cut off parallel to the cone bottom.
```

File: src/nomad_material_processing/__init__.py [hunk 11]
```diff
@@ -123,6 +244,154 @@ class Parallelepiped(Geometry):
...
+            component='NumberEditQuantity',
+            defaultDisplayUnit='millimeter ** 2',
+        ),
+        unit='meter ** 2',
+    )
+
+
+class Cylinder(Geometry):
+    """
+    Class autogenerated from yaml schema.
```
Comment (hampusnasstrom) [inline src/nomad_material_processing/__init__.py:312]:
```suggestion
    A cylinder, i.e. a prism with a circular base.
```

File: src/nomad_material_processing/__init__.py [hunk 12]
```diff
@@ -219,12 +546,12 @@ class Substrate(CompositeSystem):
     """
 
     m_def = Section()
+
     supplier = Quantity(
         type=str,
-        description='The supplier of the current substrate specimen.',
+        description='The manufacturer of the current substrate specimen.',
```
Comment (hampusnasstrom) [inline src/nomad_material_processing/__init__.py:552]:
```suggestion
        description='The supplier of the current substrate specimen.',
```

**Review by hampusnasstrom — APPROVED** [review]
Just a few small things, after that it looks great!

**Pull request merged**